cmd/snap-update-ns/change_test.go

```diff
@@ -2424,3 +2431,80 @@ func (s *changeSuite) TestPerformedChangesAreTracked(c *C) {
 		{Action: update.Keep, Entry: osutil.MountEntry{Name: "tmpfs", Dir: "/target", Type: "tmpfs"}},
 	})
 }
+
+func (s *changeSuite) TestComplexPropagatingChanges(c *C) {
+	// This problem is more subtle. It is a variant of the regression test
+	// implemented in tests/regression/lp-1831010. Here, we have four directories:
+	//
+	// - $SNAP/a
+	// - $SNAP/b
+	// - $SNAP/b/c
+	// - $SNAP/d
+	//
+	// The mount profile contains but two entries:
+	//
+	// 1) recursive-bind $SNAP/b/c -> $SNAP/a
+	// 2) recursive-bind $SNAP/d -> $SNAP/b
+	//
+	// NB: The notation "foo -> bar" can be read as "foo is a view of bar",
+	// and can be implemented by mount "--rbind bar foo".
+	//
+	// Both mount operations are performed under a substrate that is MS_SHARED.
+	// Therefore, due to the rules that decide upon propagation of bind mounts
+	// the propagation of the new mount entries is also shared. This is
+	// documented in https://lwn.net/Articles/159092/ (section 5b).
```

Comment:
Please use the official documentation, even though it is more unwieldy: https://www.kernel.org/doc/Documentation/filesystems/sharedsubtree.txt (me suggested change below adjusts for this).